lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -832,6 +1012,49 @@ std::string SPIRVToOCLBase::getGroupBuiltinPrefix(CallInst *CI) {
   return Prefix;
 }
 
+std::string
+SPIRVToOCLBase::getOCLImageOpaqueType(SmallVector<std::string, 8> &Postfixes) {
+  SmallVector<int, 7> Ops;
+  for (unsigned I = 1; I < 8; ++I)
+    Ops.push_back(atoi(Postfixes[I].c_str()));
+  SPIRVTypeImageDescriptor Desc(static_cast<SPIRVImageDimKind>(Ops[0]), Ops[1],
+                                Ops[2], Ops[3], Ops[4], Ops[5]);
+
+  std::string OCLStructName =
+      std::string(kSPR2TypeName::OCLPrefix) + rmap<std::string>(Desc);
+
+  SPIRVAccessQualifierKind Acc = static_cast<SPIRVAccessQualifierKind>(Ops[6]);
+  insertImageNameAccessQualifier(Acc, OCLStructName);
+  return OCLStructName;
+}
+
+void SPIRVToOCLBase::translateOpaqueTypes() {
+  for (auto *S : M->getIdentifiedStructTypes()) {
+    StringRef STName = S->getStructName();
+    bool IsSPIRVOpaque = STName.startswith(kSPIRVTypeName::PrefixAndDelim);
+
+    if (!IsSPIRVOpaque)
```

Comment:
Should we also check that the type referenced by `S` is an LLVM opaque type? https://llvm.org/doxygen/classllvm_1_1StructType.html#aec7c3cecd0559788b36c46df1b2181c1

Reply:
Good idea. Done :)